library/init/data/nat/basic.lean

```diff
@@ -32,8 +32,8 @@ protected def sub : ℕ → ℕ → ℕ
 | a (b+1) := pred (sub a b)
 
 protected def mul : nat → nat → nat
-| a 0     := 0
-| a (b+1) := (mul a b) + a
+| 0 b     := 0
+| (a+1) b := (mul a b) + b
```

Comment:
The term-mode expression for this is much uglier:
```
def mul._main : ℕ → ℕ → ℕ :=
λ (ᾰ ᾰ_1 : ℕ),
  ᾰ.brec_on
    (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ) (ᾰ_1 : ℕ),
       (λ (ᾰ ᾰ_1 : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ),
          ᾰ.cases_on (λ (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) 0), id_rhs ℕ 0)
            (λ (ᾰ_1_1 : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ_1_1.succ),
               id_rhs ℕ (_F.fst.fst ᾰ_1 + ᾰ_1))
            _F)
         ᾰ
         ᾰ_1
         _F)
    ᾰ_1
```
vs
```
def mul'._main : ℕ → ℕ → ℕ :=
λ (ᾰ ᾰ_1 : ℕ), ᾰ_1.cases_on (id_rhs ℕ 0) (λ (ᾰ_1 : ℕ), id_rhs ℕ (mul ᾰ ᾰ_1 + ᾰ))
```

Reply:
It can probably be rewritten to factor out the `b`

Reply:
Does
```
| 0     := λ b, 0
| (a+1) := λ b, (mul a b) + b
```
help?

Reply:
Eric, I think your recursive definition looks nicer because you accidentally referenced `mul` and not `mul'`. Here is my experiment:
```
import init.data.nat.basic

def mul_left : ℕ → ℕ → ℕ
| 0       m := 0
| (n + 1) m := mul_left n m + m

def mul_left' : ℕ → ℕ → ℕ
| 0       := λ _, 0
| (n + 1) := λ m, mul_left' n m + m

def mul_leftr : ℕ → ℕ → ℕ :=
λ n, nat.rec_on n (λ _, 0) (λ _ f m, f m + m)

def mul_right : ℕ → ℕ → ℕ
| n 0       := 0
| n (m + 1) := mul_right n m + n

def mul_right' (n : ℕ) : ℕ → ℕ
| 0       := 0
| (m + 1) := mul_right' m + n

def mul_rightr : ℕ → ℕ → ℕ :=
λ n m, nat.rec_on m 0 (λ _ k, k + n)

/-
def mul_left._main : ℕ → ℕ → ℕ :=
λ (ᾰ ᾰ_1 : ℕ),
  ᾰ.brec_on
    (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ) (ᾰ_1 : ℕ),
       (λ (ᾰ ᾰ_1 : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ),
          ᾰ.cases_on (λ (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) 0), id_rhs ℕ 0)
            (λ (ᾰ_1_1 : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ_1_1.succ),
               id_rhs ℕ (_F.fst.fst ᾰ_1 + ᾰ_1))
            _F)
         ᾰ
         ᾰ_1
         _F)
    ᾰ_1
-/

#print mul_left._main
/-
def mul_left'._main : ℕ → ℕ → ℕ :=
λ (ᾰ : ℕ),
  ᾰ.brec_on
    (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ),
       (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ),
          ᾰ.cases_on (λ (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) 0), id_rhs (ℕ → ℕ) (λ (_x : ℕ), 0))
            (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ.succ),
               id_rhs (ℕ → ℕ) (λ (m : ℕ), _F.fst.fst m + m))
            _F)
         ᾰ
         _F)
-/

#print mul_left'._main
/-
def mul_leftr : ℕ → ℕ → ℕ :=
λ (n : ℕ), n.rec_on (λ (_x : ℕ), 0) (λ (_x : ℕ) (f : ℕ → ℕ) (m : ℕ), f m + m)
-/

#print mul_leftr
/-
def mul_right._main : ℕ → ℕ → ℕ :=
λ (ᾰ ᾰ_1 : ℕ),
  ᾰ_1.brec_on
    (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ) (ᾰ_1 : ℕ),
       (λ (ᾰ ᾰ_1 : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ_1),
          ᾰ_1.cases_on (λ (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) 0), id_rhs ℕ 0)
            (λ (ᾰ_1 : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ → ℕ) ᾰ_1.succ),
               id_rhs ℕ (_F.fst.fst ᾰ + ᾰ))
            _F)
         ᾰ_1
         ᾰ
         _F)
    ᾰ
-/

#print mul_right._main
/-
def mul_right'._main : ℕ → ℕ → ℕ :=
λ (n ᾰ : ℕ),
  ᾰ.brec_on
    (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ) ᾰ),
       (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ) ᾰ),
          ᾰ.cases_on (λ (_F : nat.below (λ (ᾰ : ℕ), ℕ) 0), id_rhs ℕ 0)
            (λ (ᾰ : ℕ) (_F : nat.below (λ (ᾰ : ℕ), ℕ) ᾰ.succ), id_rhs ℕ (_F.fst.fst + n))
            _F)
         ᾰ
         _F)
-/

#print mul_right'._main
/-
def mul_rightr : ℕ → ℕ → ℕ :=
λ (n m : ℕ), m.rec_on 0 (λ (_x k : ℕ), k + n)
-/

#print mul_rightr
```